### ISSUE TITLE
send_file() function tries to get the file from filesystem even if file-like object is passed

### ISSUE THREAD

**Issue by margru** [pallets/flask#1782]
I use `send_file()` function for serving file-like objects from GridFS. However, if the file-like object obtained from the GridFS has assigned a file name, the `send_file()` function tries to get the file from the file system and I get a "File not found"-like error. It should serve the data from the file-like object directly and not to try to find the file on the file system (even when there is a file name assigned to the file-like object). Or am I missing something?

**Comment by untitaker** [comment]
The handle is used for reading the file content, but additional information is guessed off the filename if one is existent:

* The mimetype is guessed. This would work even if the filepath is pointing to a nonexistent file.
* The content length is inferred. This is actually fairly important, without content length header you basically get no percentage progress when downloading files.
* X-Sendfile support

Without a full traceback it's hard to guess which one is causing the problem.

**Comment by margru** [comment]
The traceback is as follows (I just shortened/replaced the real file paths):
```
u'/absolute/path/to/file.ext'
Traceback (most recent call last):
  File "...\lib\site-packages\flask\app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "...\lib\site-packages\flask\app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "...\blueprints\api_blueprint.py", line 1124, in req_file
    return send_file(file_obj, as_attachment=False, mimetype=file_obj.content_type)
  File "...\lib\site-packages\flask\helpers.py", line 542, in send_file
    os.path.getmtime(filename),
  File "...\lib\genericpath.py", line 54, in getmtime
    return os.stat(filename).st_mtime
WindowsError: [Error 2] Cannot find file: u'/absolute/path/to/file.ext'
```

The `file_obj` passed to the `send_file()` function on line 1124 is created as follows:
```
file_obj = grid_fs_client.get(file_id)
```

If the file is saved in the GridFS (and then loaded from the GridFS) with `.name` and `.filename` attributes set to `None` then there is no error. But if the file is saved/loaded with these attributes set to some string (e.g. `file.ext` in this case) then the described error occurs as in my case the file name is coming from the client and thus the file doesn't really exist on the server's file system. In addition, the absolute path is somehow created by the `send_file()` function.

For now, I did a workaround and I'm setting these attributes to `None` when saving files to GridFS and everything works OK but I'm loosing the file name (which was provided by the client).

I think that when the file-like object is passed as an input parameter, no operations requiring the file to exist on the file system should be performed. In addition, it could be also dangerous as if the file with the file name existed on the server's file system, it wouldn't necessarily have to be the same file which the file-like object represents.

I'm using Flask version: 0.10.1.

**Comment by untitaker** [comment]
I'd happily accept patches that make `send_file` more robust against `.name` and `.filename` properties that don't exist on the local FS, but since expecting files that are passed to `send_file` to exist on the path they are associated with is IMO a reasonable assumption I'll mark this as enhancement.

**Comment by g3wanghc** [comment]
@untitaker was wondering if adding an optional flag is a good idea to indicated that a file-like object is being passed.

**Comment by untitaker** [comment]
I don't think that's a good idea, I'd rather silently swallow the errors that occur when the `.name` and `.filename` params are invalid.

**Comment by margru** [comment]
I agree with @untitaker as it doesn't make sense to pass another flag indicating file-like object if a file-like object is indeed passed into `send_file()` function. I think it would be better to silently ignore the errors if a file-like object is passed and the associated filename doesn't exist on the FS.

**Comment by g3wanghc** [comment]
On a side note, do you guys think `sys.getsizeof` is a fair estimate of the size of file-like objects?

**Comment by RonnyPfannschmidt** [comment]
no - it would be pure accident

**Comment by ThiefMaster** [comment]
No, that function is about memory, not about file sizes

**Comment by g3wanghc** [comment]
Alright. 😭

**Comment by davidism** [comment]
Fixed in #1988.
